src/main/java/org/opensearch/jobscheduler/utils/JobDetailsService.java

```diff
@@ -0,0 +1,306 @@
+
+    public static Long TIME_OUT_FOR_LATCH = 5L;
+    private final Client client;
+    private final ClusterService clusterService;
+
+    public JobDetailsService(final Client client, final ClusterService clusterService) {
+        this.client = client;
+        this.clusterService = clusterService;
+    }
+
+    public boolean jobDetailsIndexExist() {
+        return clusterService.state().routingTable().hasIndex(JOB_DETAILS_INDEX_NAME);
+    }
+
+    /**
+     *
+     * @param listener an {@code ActionListener} that has onResponse and onFailure that is used to return the job details index if it was created
+     *      *                 or else null.
+     */
+    @VisibleForTesting
+    void createJobDetailsIndex(ActionListener<Boolean> listener) {
+        if (jobDetailsIndexExist()) {
+            listener.onResponse(true);
+        } else {
+            CreateIndexRequest request = new CreateIndexRequest(JOB_DETAILS_INDEX_NAME).mapping(jobDetailsMapping());
+            client.admin()
+                .indices()
+                .create(request, ActionListener.wrap(response -> listener.onResponse(response.isAcknowledged()), exception -> {
+                    if (exception instanceof ResourceAlreadyExistsException
+                        || exception.getCause() instanceof ResourceAlreadyExistsException) {
+                        listener.onResponse(true);
+                    } else {
+                        listener.onFailure(exception);
+                    }
+                }));
+        }
+    }
+
+    /**
+     * Attempts to process job details with a specific extension Id. If the job details does not exist it attempts to create the job details document.
+     * If the job details document exists, it will try to update the job details.
+     *
+     * @param jobIndexName a non-null job index name.
+     * @param jobTypeName a non-null job type name.
+     * @param jobParameterActionName a non-null job parameter action name.
+     * @param jobRunnerActionName a non-null job runner action name.
+     * @param extensionId the unique Id for the job details.
+     * @param listener an {@code ActionListener} that has onResponse and onFailure that is used to return the job details if it was processed
+     *                 or else null.
+     */
+    public void processJobDetailsForExtensionId(
+        final String jobIndexName,
+        final String jobTypeName,
+        final String jobParameterActionName,
+        final String jobRunnerActionName,
+        final String extensionId,
+        final JobDetailsRequestType requestType,
+        ActionListener<JobDetails> listener
+    ) {
+        boolean isJobIndexRequest;
+        if (requestType.JOB_INDEX == requestType) {
+            isJobIndexRequest = true;
+            if (jobIndexName == null) {
+                listener.onFailure(new IllegalArgumentException("Job Index Name should not be null"));
+            } else if (jobParameterActionName == null) {
```

Comment:
Since none of these can be null, seems we should include all of them in one exception.  So a single if check whether any are null, and an IAE listing which ones are null.

Reply:
done